R/qenv-join.R

```diff
@@ -23,30 +28,54 @@
 #' qq <- join(q1, q2)
 #' get_code(qq)
 #' @export
-setGeneric("join", function(x, y) {
+setGeneric("join", function(x, y, force, verbose) {
   standardGeneric("join")
 })
 
 #' @rdname join
 #' @export
-setMethod("join", signature = c("qenv", "qenv"), function(x, y) {
+setMethod("join", signature = c("qenv", "qenv"), function(x, y, force, verbose) {
+  if (missing(force) || !is.logical(force)) force <- FALSE
+  if (missing(verbose) || !is.logical(verbose)) verbose <- TRUE
+
   join_validation <- .check_joinable(x, y)
 
-  # join expressions
-  if (!isTRUE(join_validation)) {
-    stop(join_validation)
-  }
+  if (isFALSE(force)) {
+    # Default
 
-  id_unique <- !y@id %in% x@id
-  x@id <- c(x@id, y@id[id_unique])
-  x@code <- c(x@code, y@code[id_unique])
-  x@warnings <- c(x@warnings, y@warnings[id_unique])
-  x@messages <- c(x@messages, y@messages[id_unique])
+    # join expressions validation
+    if (!isTRUE(join_validation)) {
+      stop(join_validation)
+    }
 
-  # insert (and overwrite) objects from y to x
-  x@env <- rlang::env_clone(x@env, parent = parent.env(.GlobalEnv))
-  rlang::env_coalesce(env = x@env, from = y@env)
-  x
+    id_unique <- !y@id %in% x@id
+    x@id <- c(x@id, y@id[id_unique])
+    x@code <- c(x@code, y@code[id_unique])
+    x@warnings <- c(x@warnings, y@warnings[id_unique])
+    x@messages <- c(x@messages, y@messages[id_unique])
+
+    # insert (and overwrite) objects from y to x
+    x@env <- rlang::env_clone(x@env, parent = parent.env(.GlobalEnv))
+    rlang::env_coalesce(env = x@env, from = y@env)
+    x
+  } else {
+    if (verbose) {
+      message("The join method of qenv was forced, please be careful and manually validate the results.")
+      if (isTRUE(join_validation)) {
+        warning("Forced join is not recomended for this scenario, the default mechanism seems to be suited.")
```

Comment:
So I don't like the fact that join(x, y, force = TRUE) and join(x, y, force = FALSE) give different answers if check joinable is true (and both qenv's have some shared code at the beginning)  as that's confusing -> we could instead change it so force = TRUE only changes the behaviour if it would have given an error.

Or better we could use a different function like concatenate or just `c` as that's what force join is effectively doing

Reply:
For me the default behaviour is more surprising than the forced one.

"we could instead change it so force = TRUE only changes the behaviour if it would have given an error."
My idea about force it is that it works always and at the same predicted way. 
We simply join what is provided.